theano/tensor/nnet/tests/test_spatialtf.py

```diff
@@ -0,0 +1,226 @@
+        y_t = T.dot(_linspace(-1.0, 1.0, height).dimshuffle(0, 'x'),
+                    T.ones((1, width)))
+
+        x_t_flat = x_t.reshape((1, -1))
+        y_t_flat = y_t.reshape((1, -1))
+        ones = T.ones_like(x_t_flat)
+        grid = T.concatenate([x_t_flat, y_t_flat, ones], axis=0)
+        return grid
+
+    num_batch, num_channels, height, width = inp.shape
+    theta = T.reshape(theta, (-1, 2, 3))
+
+    # grid of (x_t, y_t, 1), eq (1) in ref [1]
+    out_height = T.cast(T.ceil(height * scale_height), 'int64')
+    out_width = T.cast(T.ceil(width * scale_width), 'int64')
+    grid = _meshgrid(out_height, out_width)
+    # transform a x (x_t, y_t, 1)^t -> (x_s, y_s)
+    t_g = T.dot(theta, grid)
+    x_s = t_g[:, 0]
+    y_s = t_g[:, 1]
+    x_s_flat = x_s.flatten()
+    y_s_flat = y_s.flatten()
+
+    # dimshuffle input to  (bs, height, width, channels)
+    input_dim = inp.dimshuffle(0, 2, 3, 1)
+    input_transformed = _interpolate(
+        input_dim, x_s_flat, y_s_flat,
+        out_height, out_width, border_mode)
+
+    output = T.reshape(
+        input_transformed, (num_batch, out_height, out_width, num_channels))
+    output = output.dimshuffle(0, 3, 1, 2)  # dimshuffle to conv format
+    return output
+
+
+class TestTransformer(unittest.TestCase):
+    def setUp(self):
+        utt.seed_rng()
+
+    def getInputs(self):
+        inp_shape = (np.random.randint(1, 11), 1, 32, 32)
+        num_images, num_channels, height, width = inp_shape
+
+        inp = np.random.random(inp_shape).astype(config.floatX)
+
+        transform = [[-1, 0, 0],
+                     [0, -1, 0]]
+        theta = np.asarray(num_images * [transform], dtype=config.floatX)
+        scale_height = np.random.random()
+        scale_width = np.random.random()
+
+        return (inp, theta, scale_width, scale_height)
+
+    def test_symb(self):
+        """
+        Compare CPU implementation with a symbolic one
+        """
+        inp, theta, scale_width, scale_height = self.getInputs()
+
+        # Setup symbolic variables
+        t_inp = T.tensor4('inp')
+        t_theta = T.tensor3('theta')
+        t_scale_height = T.scalar('scale_height')
+        t_scale_width = T.scalar('scalar_width')
+
+        symb_out_op = spatialtf_symb(t_inp, t_theta, t_scale_width, t_scale_height)
+        symb_out_fn = theano.function([t_inp, t_theta, t_scale_width, t_scale_height],
+                                      symb_out_op, mode=mode_without_gpu)
+        symb_out = symb_out_fn(inp, theta, scale_width, scale_height)
+
+        cpu_out_op = spatialtf(t_inp, t_theta, t_scale_width, t_scale_height)
+        cpu_out_fn = theano.function([t_inp, t_theta, t_scale_width, t_scale_height],
+                                     cpu_out_op, mode=mode_without_gpu)
+        cpu_out = cpu_out_fn(inp, theta, scale_width, scale_height)
+
+        # Check if results are approx. equal
+        utt.assert_allclose(symb_out, cpu_out)
```

Comment:
I don't think the purpose of symbolic implementation should be to compare its results with your numpy-based implementation, as it is quite useless. To test your numpy-based implementation, you should better compare results with expected hardcoded concrete values, just as you did to test CPU CTC implementation.

However, about symbolic implementation, it may be useful to compare it to numpy-based implementation for benchmarking, to check which version is the fastest (in practice, symbolic implementation might be fastest depending on Theano implementations). But a benchmark is not a unit test. So I suggest to keep the symbolic implementation, remove this test (`test_symb`), replace it with a real test (which compares your values with expected hardcoded values), and manually benchmark both implementstions yourself (especially with big inputs).

Reply:
I agree, but I think, however, the highest priority is getting a complete functional spatial transformer. As soon as I fix the gradient of the transformation and improve the tests, translating NumPy code into Theano symbolic operations is straightforward.

Reply:
Indeed, let's finish and fix the gradient implementation, and then we will look for ultimate corrections.